onecodex/helpers.py

```diff
@@ -1,169 +1,328 @@
+                            if str_f.lower() == tax_name.lower():
+                                hits = [(tax_id, tax_name)]
+                                break
+                            # otherwise, keep trying to match
+                            elif str_f.lower() in tax_name.lower():
+                                hits.append((tax_id, tax_name))
+
+                        # take the hit with the lowest tax_id
+                        hits = sorted(hits, key=lambda x: int(x[0]))
+
+                    if hits:
+                        # report within-rank abundance
+                        df = self.results(rank=self.taxonomy['rank'][hits[0][0]])
+
+                        renamed_field = '{} ({})'.format(hits[0][1], hits[0][0])
+                        magic_metadata[renamed_field] = df[hits[0][0]]
+                        magic_fields[f] = renamed_field
+                    else:
+                        # matched nothing
+                        raise OneCodexException(
+                            'Field or taxon {} not found. Choose from: {}'.format(str_f, help_metadata)
+                        )
+
+        return magic_metadata, magic_fields
+
+    def results(self, rank='auto', top_n=None, threshold=None,
+                remove_zeros=True, normalize='auto',
+                table_format='wide'):
+        """Takes the ResultsDataFrame associated with these samples, or SampleCollection, does some
+        filtering, and returns a ResultsDataFrame.
+
+        Parameters
+        ----------
+        rank : {'auto', 'kingdom', 'phylum', 'class', 'order', 'family', 'genus', 'species'}, optional
+            Analysis will be restricted to abundances of taxa at the specified level.
+        top_n : `integer`, optional
+            Return only the top N most abundant taxa.
+        threshold : `float`, optional
+            Return only taxa more abundant than this threshold in one or more samples.
+        remove_zeros : `bool`, optional
+            Do not return taxa that have zero abundance in every sample.
+        normalize : {'auto', True, False}
+            Convert read counts to relative abundances (each sample sums to 1.0).
+        table_format : {'long', 'wide'}
+            If wide, rows are classifications, cols are taxa, elements are counts. If long, rows are
+            observations with three cols each: classification_id, tax_id, and count.
+
+        Returns
+        -------
+        ResultsDataFrame
+        """
+
+        rank = self._get_auto_rank(rank)
+        df = self._results.copy()
+
+        # subset by taxa
+        if rank:
+            tax_ids_to_keep = []
+
+            for tax_id in df.keys():
+                if self.taxonomy['rank'][tax_id] == rank:
+                    tax_ids_to_keep.append(tax_id)
+
+            if len(tax_ids_to_keep) == 0:
+                raise OneCodexException('No taxa kept--is rank ({}) correct?'.format(rank))
```

Comment:
Are the rank names stored anywhere? It might be better to assert that the rank is valid early.